### PR TITLE
Dev

### PR DESCRIPTION
**Pull request by cfrancesco**
Introduced Forward Transformer and it's pipeline. A few fixes. Heavy refactoring. Doc and README updates.

### CODE REVIEW

File: model/models.py [hunk 1]
```diff
@@ -239,18 +257,217 @@ def load_weights(self, weights_path: str, r: int = 1):
...
+        if drop_n_heads is not None:
+            self.drop_n_heads = drop_n_heads
+    
+    def __apply_signature(self, function, signature):
+        if self.debug:
+            return function
+        else:
+            return tf.function(input_signature=signature)(function)
+    
+    def encode_text(self, text):
```
Comment (cschaefer26) [inline model/models.py:446]: Mixed order of private and public methods.

File: utils/alignments.py [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+
+def binary_attention(attention_weights):
+    attention_peak_per_phoneme = attention_weights.max(axis=1)
+    binary_attn = (attention_weights.T == attention_peak_per_phoneme).astype(int).T
+    binary_score = np.sum(attention_weights * binary_attn)
+    return binary_attn, binary_score
+
+
+def get_durations_from_alignment(batch_alignments, mels, phonemes, weighted=True, PLOT_OUTLIERS=False, PLOT_ALL=False):
+    mel_pad_mask = create_mel_padding_mask(mels)
```
Comment (cschaefer26) [inline utils/alignments.py:92]: This is pretty compliated, I think it would be beneficial to have some comments.

File: utils/config_loader.py [hunk 3]
```diff
@@ -1,52 +1,120 @@
...
+        base_dir = Path(self.config['log_directory']) / self.session_name
+        log_dir = base_dir / f'{self.model_kind}_logs'
+        weights_dir = base_dir / f'{self.model_kind}_weights'
+        train_datadir = self.config['train_data_directory']
+        if train_datadir is None:
+            train_datadir = self.config['data_directory']
+        train_datadir = Path(train_datadir)
+        return base_dir, log_dir, train_datadir, weights_dir
+    
+    def _check_hash(self):
```
Comment (cschaefer26) [inline utils/config_loader.py:62]: Mixed order of private and public methods.

File: config/standard/autoregressive_config.yaml [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+postnet_conv_layers: 5
+postnet_kernel_size: 5
+encoder_dense_blocks: 4
+decoder_dense_blocks: 4
+
+# LOSSES
+stop_loss_scaling: 8
+
+# TRAINING
+dropout_rate: 0.1
```
[the remark this answers is not in the document]
Comment (cfrancesco) [reply config/standard/autoregressive_config.yaml:21]: we set dropout rate to be the training dropout for all the layers. the scheduled dropout is now called decoder_prenet_dropout and in the config is only defined by the schedule

**Comment by datitran** [comment]
As discussed I think it's better to split the PR into two: one PR should contain the refactoring of code, small fixes and a clean README for the TransformerTTS. The second should be the added ForwardTransformer Code + Model.

File: config/standard/forward_config.yaml [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+debug: False
+
+# LOGGING
+validation_frequency: 1_000
+prediction_frequency: 1_000
+weights_save_frequency: 5_000
+train_images_plotting_frequency: 1_000
+keep_n_weights: 5
+keep_checkpoint_every_n_hours: 12
+n_steps_avg_losses: [100, 500, 1_000, 5_000]
```
[the remark this answers is not in the document]
Comment (cfrancesco) [reply config/standard/forward_config.yaml:45]: the number of steps? it's a CLI display feature that shows the avg loss over the last n steps. useful to get a quick understanding if the model is learning, plateauing and or overfitting

File: model/layers.py [hunk 6]
```diff
@@ -158,80 +226,115 @@ def __init__(self, model_dim: int, num_heads: int, dropout_rate: float = 0.1, **
...
+            CrossAttentionDenseBlock(model_dim=model_dim, dropout_rate=dropout_rate, num_heads=n_heads,
+                                     dense_hidden_units=feed_forward_dimension, name=f'{self.name}_CADB_{i}')
+            for i, n_heads in enumerate(num_heads[:dense_blocks])]
+        self.CACB = [
+            CrossAttentionConvBlock(model_dim=model_dim, dropout_rate=dropout_rate, num_heads=n_heads,
+                                    name=f'{self.name}_CACB_{i}')
+            for i, n_heads in enumerate(num_heads[dense_blocks:])]
     
-    def call(self, inputs, enc_output, training, look_ahead_mask, padding_mask):
+    def call(self, inputs, enc_output, training, decoder_padding_mask, encoder_padding_mask, drop_n_heads,
```
[the remark this answers is not in the document]
Comment (cfrancesco) [reply model/layers.py:303]: I don't understand. Maybe worth to point out that the reduction factor is a model class attribute

File: requirements.txt [hunk 7]
```diff
@@ -1,6 +1,7 @@
+matplotlib
 librosa==0.7.1
 numpy==1.17.4
 phonemizer==2.1
 ruamel.yaml==0.16.6
-tensorflow==2.1.0
+tensorflow==2.2.0
```
[the remark this answers is not in the document]
Comment (cfrancesco) [reply requirements.txt:6]: Yeah, I relaxed other packages too. Should be >=2.1 (but also <=2.2 :D)

File: train_forward.py [hunk 8]
```diff
@@ -0,0 +1,218 @@
...
+        for gpu in gpus:
+            tf.config.experimental.set_memory_growth(gpu, True)
+        logical_gpus = tf.config.experimental.list_logical_devices('GPU')
+        print(len(gpus), "Physical GPUs,", len(logical_gpus), "Logical GPUs")
+    except RuntimeError as e:
+        # Memory growth must be set before GPUs have been initialized
+        print(e)
+
+
+# aux functions declaration
```
[the remark this answers is not in the document]
Comment (cfrancesco) [reply train_forward.py:34]: that still depends on training flags, with next refactor I'll move it too to config loader

**Comment by cfrancesco** [comment]
Ok, I split the PR
